In [1]:
import pandas as pd

In [14]:
dx = pd.read_csv('Student_Marks.csv') 
#akses di level folder yg di akses
dx

number_courses  time_study   Marks
0                3       4.508  19.202
1                4       0.096   7.734
2                4       3.133  13.811
3                6       7.909  53.018
4                8       7.811  55.299
..             ...         ...     ...
95               6       3.561  19.128
96               3       0.301   5.609
97               4       7.163  41.444
98               7       0.309  12.027
99               3       6.335  32.357

[100 rows x 3 columns]

# jika 
file tsb di dalam folder lain maka namafolder/namafile, 
kalo mau naik folder atau filenya di luar folder sekarang maka, harus keluar folder terus masuk ke data seet untuk akses student mark

## caranya
('../foldertujuan/filetujuan.csv')
.. disini naik 1 level
('../../foldertujuan/filetujuan.csv')
.. disini naik 2 level

In [3]:
pd.read_csv('../Student_Marks.csv')
#akses 1 level di atas folder di akses

number_courses  time_study   Marks
0                3       4.508  19.202
1                4       0.096   7.734
2                4       3.133  13.811
3                6       7.909  53.018
4                8       7.811  55.299
..             ...         ...     ...
95               6       3.561  19.128
96               3       0.301   5.609
97               4       7.163  41.444
98               7       0.309  12.027
99               3       6.335  32.357

[100 rows x 3 columns]

In [4]:
pd.read_csv('../../Student_Marks.csv')
#akses 2 level di atas folder di akses

number_courses  time_study   Marks
0                3       4.508  19.202
1                4       0.096   7.734
2                4       3.133  13.811
3                6       7.909  53.018
4                8       7.811  55.299
..             ...         ...     ...
95               6       3.561  19.128
96               3       0.301   5.609
97               4       7.163  41.444
98               7       0.309  12.027
99               3       6.335  32.357

[100 rows x 3 columns]

In [8]:
pd.read_csv('ref.csv')

ch1;ch2;ch3
0    10;1;100
1    20;2;200
2    30;3;300
3    40;4;400
4    50;5;500

In [10]:
pd.read_csv('ref.csv', sep=';')
#disini kita pakai pemisah di rubah

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500

In [13]:
pd.read_csv('ref.csv', delimiter =';')
#disini kita pakai pemisah di rubah kalau pemisahnya TAB jadi kita pakai /t

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500


# disini kita pakai pemisah di rubah kalau pemisahnya TAB jadi kita pakai /t

In [16]:
df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')
df

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
0                1            0     TRH    Huskies  ...  40.294830     NYK   
1                1            0     NYK     Knicks  ...  41.705170     TRH   
2                1            0     CHS      Stags  ...  42.012257     NYK   
3                2            0     NYK     Knicks  ...  40.692783     CHS   
4                1            0     DTF    Falcons  ...  38.864048     WSC   
...            ...          ...     ...        ...  ...        ...     ...   
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
0          Knicks       68  1300.0000  1306.7233              H            L   
1         Huskies       66  1300.0000  1293.2767              A            W   
2          Knicks       47  1306.7233  1297.0712              H            W   
3           Stags       63  1300.0000  1309.6521              A            L   
4        Capitols       50  1300.0000  1320.3811              H            L   
...           ...      ...        ...        ...            ...          ...   
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
0       0.640065   NaN  
1       0.359935   NaN  
2       0.631101   NaN  
3       0.368899   NaN  
4       0.640065   NaN  
...          ...   ...  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[126314 rows x 23 columns]

In [17]:
len(df)

126314

In [18]:
df.shape

(126314, 23)

In [20]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [21]:
df.head

<bound method NDFrame.head of         gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
0                1            0     TRH    Huskies 

In [22]:
pd.set_option('display.max.columns', None)

In [24]:
df.head() # ini nampilin 5 data teratas

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

In [25]:
df.tail() #ini view 5 data terbawah

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126310         102            1     GSW   Warriors  104  1809.9791  1813.6349   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   
126313         103            1     GSW   Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309  60.309792    GSW   Warriors      103  1790.9591  1809.9791   
126310  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126311  60.010067    GSW   Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW   Warriors      105  1813.6349  1822.2881   
126313  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
126309             H           L  0.546572   NaN  
126310             H           W  0.765565   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  
126313             A           W  0.518550   NaN

## Data Overview
    ini isi kata" biar kamo bisa di baca

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

In [30]:
df.describe()

gameorder        _iscopy        year_id     seasongame  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean    31579.000000       0.500000    1988.200374      43.533733   
std     18231.927643       0.500002      17.582309      25.375178   
min         1.000000       0.000000    1947.000000       1.000000   
25%     15790.000000       0.000000    1975.000000      22.000000   
50%     31579.000000       0.500000    1990.000000      43.000000   
75%     47368.000000       1.000000    2003.000000      65.000000   
max     63157.000000       1.000000    2015.000000     108.000000   

         is_playoffs            pts          elo_i          elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean        0.063857     102.729982    1495.236055    1495.236055   
std         0.244499      14.814845     112.139945     112.461687   
min         0.000000       0.000000    1091.644500    1085.774400   
25%         0.000000      93.000000    1417.237975    1416.994900   
50%         0.000000     103.000000    1500.945550    1500.954400   
75%         0.000000     112.000000    1576.060000    1576.291625   
max         1.000000     186.000000    1853.104500    1853.104500   

           win_equiv        opp_pts      opp_elo_i      opp_elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean       41.707889     102.729982    1495.236055    1495.236055   
std        10.627332      14.814845     112.139945     112.461687   
min        10.152501       0.000000    1091.644500    1085.774400   
25%        34.103035      93.000000    1417.237975    1416.994900   
50%        42.113357     103.000000    1500.945550    1500.954400   
75%        49.635328     112.000000    1576.060000    1576.291625   
max        71.112038     186.000000    1853.104500    1853.104500   

            forecast  
count  126314.000000  
mean        0.500000  
std         0.215252  
min         0.020447  
25%         0.327989  
50%         0.500000  
75%         0.672011  
max         0.979553

In [29]:
df.describe().T # merubah dr kesamping ke bawah

count          mean           std          min           25%  \
gameorder    126314.0  31579.000000  18231.927643     1.000000  15790.000000   
_iscopy      126314.0      0.500000      0.500002     0.000000      0.000000   
year_id      126314.0   1988.200374     17.582309  1947.000000   1975.000000   
seasongame   126314.0     43.533733     25.375178     1.000000     22.000000   
is_playoffs  126314.0      0.063857      0.244499     0.000000      0.000000   
pts          126314.0    102.729982     14.814845     0.000000     93.000000   
elo_i        126314.0   1495.236055    112.139945  1091.644500   1417.237975   
elo_n        126314.0   1495.236055    112.461687  1085.774400   1416.994900   
win_equiv    126314.0     41.707889     10.627332    10.152501     34.103035   
opp_pts      126314.0    102.729982     14.814845     0.000000     93.000000   
opp_elo_i    126314.0   1495.236055    112.139945  1091.644500   1417.237975   
opp_elo_n    126314.0   1495.236055    112.461687  1085.774400   1416.994900   
forecast     126314.0      0.500000      0.215252     0.020447      0.327989   

                      50%           75%           max  
gameorder    31579.000000  47368.000000  63157.000000  
_iscopy          0.500000      1.000000      1.000000  
year_id       1990.000000   2003.000000   2015.000000  
seasongame      43.000000     65.000000    108.000000  
is_playoffs      0.000000      0.000000      1.000000  
pts            103.000000    112.000000    186.000000  
elo_i         1500.945550   1576.060000   1853.104500  
elo_n         1500.954400   1576.291625   1853.104500  
win_equiv       42.113357     49.635328     71.112038  
opp_pts        103.000000    112.000000    186.000000  
opp_elo_i     1500.945550   1576.060000   1853.104500  
opp_elo_n     1500.954400   1576.291625   1853.104500  
forecast         0.500000      0.672011      0.979553

untuk melihat anomali data yang ada di data sheet kita gunakan describe

## Data Exproation

In [32]:
df.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233   41.70517    TRH   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   
1  Huskies       66     1300.0  1293.2767             A           W  0.359935   

  notes  
0   NaN  
1   NaN

In [33]:
df['game_result']

0         L
1         W
2         W
3         L
4         L
         ..
126309    L
126310    W
126311    L
126312    L
126313    W
Name: game_result, Length: 126314, dtype: object

In [34]:
df['game_result'].unique()

array(['L', 'W'], dtype=object)

In [36]:
df['game_result'].value_counts()

W    63157
L    63157
Name: game_result, dtype: int64

In [41]:
df['fran_id'].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

In [45]:
df.sort_values(by='pts', ascending=False)[['fran_id','pts','year_id']]

fran_id  pts  year_id
50094  Pistons  186     1984
50095  Nuggets  184     1984
21416   Pacers  177     1970
32881    Sails  176     1975
8974   Celtics  173     1959
...        ...  ...      ...
4      Falcons   33     1947
3191   Pistons   19     1951
3190    Lakers   18     1951
26685  Squires    2     1973
26684  Nuggets    0     1973

[126314 rows x 3 columns]

ascending
descending
True
False
bisa di kombinasikan sekendak mu

In [48]:
df.head(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0  1309.6521  42.012257    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68  1300.0000  1306.7233             H           L  0.640065   
1  Huskies       66  1300.0000  1293.2767             A           W  0.359935   
2   Knicks       47  1306.7233  1297.0712             H           W  0.631101   

  notes  
0   NaN  
1   NaN  
2   NaN

In [54]:
df.loc[df['fran_id']=='Knicks','opp_fran'].value_counts()

Celtics         526
Sixers          471
Pistons         382
Hawks           368
Warriors        335
Wizards         323
Lakers          310
Kings           299
Bulls           251
Pacers          205
Bucks           204
Cavaliers       203
Nets            189
Rockets         145
Heat            137
Thunder         124
Clippers        122
Suns            120
Trailblazers    105
Magic            99
Jazz             90
Spurs            90
Pelicans         90
Raptors          85
Nuggets          81
Mavericks        68
Baltimore        62
Timberwolves     50
Hornets          40
Grizzlies        37
Capitols         32
Stags            23
Bombers          23
Olympians        20
Steamrollers     20
Rebels            9
Ironmen           6
Falcons           6
Huskies           6
Jets              5
Denver            2
Redskins          2
Waterloo          2
Packers           2
Name: opp_fran, dtype: int64

In [63]:
df.loc[(df['year_id']<=2010) & (df['fran_id']=='Knicks'), :]

#salah penulisan erorr nian la

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
1               1  194611010TRH   NBA        1     1947   11/1/1946   
3               2  194611020CHS   NBA        1     1947   11/2/1946   
23             12  194611070STB   NBA        1     1947   11/7/1946   
37             19  194611110NYK   NBA        0     1947  11/11/1946   
54             28  194611160NYK   NBA        0     1947  11/16/1946   
...           ...           ...   ...      ...      ...         ...   
113352      56677  201004070IND   NBA        1     2010    4/7/2010   
113396      56699  201004090ORL   NBA        1     2010    4/9/2010   
113428      56715  201004110NYK   NBA        0     2010   4/11/2010   
113447      56724  201004120NYK   NBA        0     2010   4/12/2010   
113486      56744  201004140TOR   NBA        1     2010   4/14/2010   

        seasongame  is_playoffs team_id fran_id  pts      elo_i      elo_n  \
1                1            0     NYK  Knicks   68  1300.0000  1306.7233   
3                2            0     NYK  Knicks   47  1306.7233  1297.0712   
23               3            0     NYK  Knicks   68  1297.0712  1307.3197   
37               4            0     NYK  Knicks   68  1307.3197  1293.2161   
54               5            0     NYK  Knicks   64  1293.2161  1296.2185   
...            ...          ...     ...     ...  ...        ...        ...   
113352          78            0     NYK  Knicks  105  1409.8129  1405.4998   
113396          79            0     NYK  Knicks  103  1405.4998  1403.5228   
113428          80            0     NYK  Knicks   98  1403.5228  1394.4866   
113447          81            0     NYK  Knicks  114  1394.4866  1400.5443   
113486          82            0     NYK  Knicks  113  1400.5443  1391.4211   

        win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
1       41.705170    TRH  Huskies       66  1300.0000  1293.2767   
3       40.692783    CHS    Stags       63  1300.0000  1309.6521   
23      41.767715    STB  Bombers       63  1312.8473  1302.5988   
37      40.288467    CHS    Stags       78  1315.2522  1329.3558   
54      40.603352    PIT  Ironmen       62  1277.5587  1274.5563   
...           ...    ...      ...      ...        ...        ...   
113352  30.973991    IND   Pacers      113  1482.3267  1486.6398   
113396  30.779263    ORL    Magic      118  1694.5193  1696.4962   
113428  29.895554    MIA     Heat      111  1580.2286  1589.2649   
113447  30.486807    WAS  Wizards      103  1352.4906  1346.4329   
113486  29.598228    TOR  Raptors      131  1424.2421  1433.3652   

       game_location game_result  forecast notes  
1                  A           W  0.359935   NaN  
3                  A           L  0.368899   NaN  
23                 A           W  0.339290   NaN  
37                 H           L  0.629479   NaN  
54                 H           W  0.660557   NaN  
...              ...         ...       ...   ...  
113352             A           L  0.270306   NaN  
113396             A           L  0.096270   NaN  
113428             H           L  0.391371   NaN  
113447             H           W  0.693683   NaN  
113486             A           L  0.329143   NaN  

[5354 rows x 23 columns]

AGREGRASI

In [64]:
df.head(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0  1309.6521  42.012257    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68  1300.0000  1306.7233             H           L  0.640065   
1  Huskies       66  1300.0000  1293.2767             A           W  0.359935   
2   Knicks       47  1306.7233  1297.0712             H           W  0.631101   

  notes  
0   NaN  
1   NaN  
2   NaN

In [66]:
df.loc[df['team_id']== 'TRH','year_id'].max()

1947

In [67]:
df.loc[df['team_id']== 'TRH','year_id'].min()

1947

team TRH hanya main 1 season saja

In [71]:
df.loc[df['team_id']== 'NYK','year_id'].agg(['min','max','mean'])

min     1947.000000
max     2015.000000
mean    1981.886289
Name: year_id, dtype: float64

team dengan team_id bermain selama season 1947 hingga 2015

In [72]:
df.sample(6)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
51351       25676  198404060CHI   NBA        0     1984   4/6/1984   
84449       42225  199903270SAS   NBA        0     1999  3/27/1999   
119943      59972  201302100SAC   NBA        0     2013  2/10/2013   
83970       41986  199902260GSW   NBA        1     1999  2/26/1999   
14536        7269  196612070BOS   NBA        0     1967  12/7/1966   
4408         2205  195203010BLB   NBA        1     1952   3/1/1952   

        seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
51351           77            0     CHI     Bulls  111  1347.1632  1357.4896   
84449           29            0     SAS     Spurs   99  1635.0851  1638.0978   
119943          52            0     SAC     Kings  117  1367.1450  1378.9485   
83970           12            0     PHO      Suns   90  1554.3102  1545.7473   
14536           23            0     BOS   Celtics  119  1640.8550  1645.6609   
4408            59            0     PHW  Warriors   77  1541.4517  1523.6512   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
51351   26.595484    ATL      Hawks      103  1483.5227  1473.1963   
84449   55.177029    DAL  Mavericks       77  1373.4998  1370.4871   
119943  28.295506    HOU    Rockets      111  1599.8416  1588.0381   
83970   46.073120    GSW   Warriors       94  1363.5004  1372.0632   
14536   57.372494    CIN      Kings       91  1431.4063  1426.6003   
4408    42.421051    BLB  Baltimore       90  1326.5334  1344.3339   

       game_location game_result  forecast notes  
51351              H           W  0.447865   NaN  
84449              H           W  0.889087   NaN  
119943             H           W  0.317808   NaN  
83970              A           L  0.627789   NaN  
14536              H           W  0.855861   NaN  
4408               A           L  0.659602   NaN

In [73]:
df.sample(2)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
19183       9592  196904040LAL   NBA        1     1969    4/4/1969   
12767       6384  196410240BOS   NBA        1     1965  10/24/1964   

       seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
19183          87            1     SFW  Warriors   98  1448.1859  1445.7010   
12767           4            0     BAL   Wizards  103  1450.9303  1445.4728   

       win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location  \
19183  41.193184    LAL   Lakers      103  1579.2096  1581.6945             A   
12767  35.639687    BOS  Celtics      131  1637.3635  1642.8210             A   

      game_result  forecast notes  
19183           L  0.209179   NaN  
12767           L  0.161265   NaN

In [76]:
df.loc[df['fran_id']=='Warriors','pts'].sum()
#untuk melihat si warriors ini mencetak berapa selama ini

591224

In [77]:
df.loc[df['fran_id']=='Warriors',['seasongame','pts']].mean()

seasongame     41.988863
pts           104.511932
dtype: float64

In [80]:
df.sample(1)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
99152      49577  200501280MIN   NBA        0     2005  1/28/2005          43   

       is_playoffs team_id       fran_id  pts      elo_i      elo_n  \
99152            0     MIN  Timberwolves   93  1548.3613  1550.1672   

       win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location  \
99152  46.562069    UTA     Jazz       91  1428.2655  1426.4596             H   

      game_result  forecast notes  
99152           W  0.780224   NaN

In [81]:
df.iloc[99152,:] #kita ngerun yg di sample atas, kta run berdasarkan index bukan label

gameorder               49577
game_id          200501280MIN
lg_id                     NBA
_iscopy                     0
year_id                  2005
date_game           1/28/2005
seasongame                 43
is_playoffs                 0
team_id                   MIN
fran_id          Timberwolves
pts                        93
elo_i                 1548.36
elo_n                 1550.17
win_equiv             46.5621
opp_id                    UTA
opp_fran                 Jazz
opp_pts                    91
opp_elo_i             1428.27
opp_elo_n             1426.46
game_location               H
game_result                 W
forecast             0.780224
notes                     NaN
Name: 99152, dtype: object

In [83]:
df.iloc[74521,1:6] # ambil index 74521 tampilkan kolom 1-5

game_id      199501060ORL
lg_id                 NBA
_iscopy                 1
year_id              1995
date_game        1/6/1995
Name: 74521, dtype: object

In [86]:
df.loc[74521:74530, ['game_id', 'lg_id','_iscopy','year_id','date_game']].min()

game_id      199501060ORL
lg_id                 NBA
_iscopy                 0
year_id              1995
date_game        1/6/1995
dtype: object

In [87]:
df.loc[74521:74530, ['game_id', 'lg_id','_iscopy','year_id','date_game']].max()

game_id      199501070HOU
lg_id                 NBA
_iscopy                 1
year_id              1995
date_game        1/7/1995
dtype: object

pas di aggregator min max dia akan ambil min max per masing" kolom

fungsi() ini yang biasanya karena fungsinya adalah pengimputan parameter
fungsi atau atribut di belakangnya pasti ada ()

sendangkan kalo [] di gunakan untuk INDEXING, pengen slice atau menampilkan kolom

tapi kembali lagi liat ke dokumentasinya

# mencari kejadian spesifik yang terjadi di daata sheet kuy


## Query

In [88]:
df.sample(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
57667       28834  198705160LAL   NBA        0     1987  5/16/1987   
18580        9291  196902140DET   NBA        1     1969  2/14/1969   
122162      61082  201401130CHI   NBA        1     2014  1/13/2014   

        seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
57667           91            1     LAL   Lakers   92  1715.6145  1717.4933   
18580           61            0     PHO     Suns  123  1223.4022  1220.9285   
122162          36            0     WAS  Wizards  102  1467.9077  1487.3230   

        win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
57667   61.461380    SEA  Thunder       87  1532.3381  1530.4594   
18580   19.859653    DET  Pistons      128  1355.2896  1357.7632   
122162  39.211369    CHI    Bulls       88  1496.5426  1477.1273   

       game_location game_result  forecast notes  
57667              H           W  0.836261   NaN  
18580              A           L  0.208358   NaN  
122162             A           W  0.322899   NaN

In [89]:
df[
    df['year_id']==2001
]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87750      43876  200010310ATL   NBA        0     2001  10/31/2000   
87751      43876  200010310ATL   NBA        1     2001  10/31/2000   
87752      43877  200010310CHI   NBA        0     2001  10/31/2000   
87753      43877  200010310CHI   NBA        1     2001  10/31/2000   
87754      43878  200010310DAL   NBA        0     2001  10/31/2000   
...          ...           ...   ...      ...      ...         ...   
90265      45133  200106100PHI   NBA        0     2001   6/10/2001   
90266      45134  200106130PHI   NBA        0     2001   6/13/2001   
90267      45134  200106130PHI   NBA        1     2001   6/13/2001   
90268      45135  200106150PHI   NBA        0     2001   6/15/2001   
90269      45135  200106150PHI   NBA        1     2001   6/15/2001   

       seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
87750           1            0     ATL      Hawks   82  1366.9685  1354.0496   
87751           1            0     CHH   Pelicans  106  1557.8643  1570.7832   
87752           1            0     CHI      Bulls   81  1326.7501  1316.6208   
87753           1            0     SAC      Kings  100  1535.6016  1545.7308   
87754           1            0     DAL  Mavericks   97  1541.3955  1545.2712   
...           ...          ...     ...        ...  ...        ...        ...   
90265         103            1     PHI     Sixers   91  1605.4132  1599.5922   
90266         104            1     PHI     Sixers   86  1599.5922  1589.4489   
90267          97            1     LAL     Lakers  100  1760.3232  1770.4666   
90268         105            1     PHI     Sixers   96  1589.4489  1581.0239   
90269          98            1     LAL     Lakers  108  1770.4666  1778.8915   

       win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87750  26.311760    CHH  Pelicans      106  1557.8643  1570.7832   
87751  48.286263    ATL     Hawks       82  1366.9685  1354.0496   
87752  23.012793    SAC     Kings      100  1535.6016  1545.7308   
87753  45.715614    CHI     Bulls       81  1326.7501  1316.6208   
87754  45.668037    MIL     Bucks       93  1522.6825  1518.8068   
...          ...    ...       ...      ...        ...        ...   
90265  51.169765    LAL    Lakers       96  1754.5022  1760.3232   
90266  50.164997    LAL    Lakers      100  1760.3232  1770.4666   
90267  65.530334    PHI    Sixers       86  1599.5922  1589.4489   
90268  49.321487    LAL    Lakers      108  1770.4666  1778.8915   
90269  66.090416    PHI    Sixers       96  1589.4489  1581.0239   

      game_location game_result  forecast notes  
87750             H           L  0.372095   NaN  
87751             A           W  0.627905   NaN  
87752             H           L  0.348282   NaN  
87753             A           W  0.651718   NaN  
87754             H           W  0.664490   NaN  
...             ...         ...       ...   ...  
90265             H           L  0.429822   NaN  
90266             H           L  0.413480   NaN  
90267             A           W  0.586520   NaN  
90268             H           L  0.385474   NaN  
90269             A           W  0.614526   NaN  

[2520 rows x 23 columns]

In [91]:
df[
    (df['year_id']==2001) & 
    (df['fran_id']=='Celtics')&
    (df['game_location']=='H')&
    (df['game_result']=='W')
]
#seperti filter pada exel BISA ALAAAAAH!!!!

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87776      43889  200011010BOS   NBA        0     2001   11/1/2000   
87805      43903  200011030BOS   NBA        0     2001   11/3/2000   
87970      43986  200011150BOS   NBA        0     2001  11/15/2000   
88001      44001  200011170BOS   NBA        0     2001  11/17/2000   
88068      44035  200011220BOS   NBA        0     2001  11/22/2000   
88093      44047  200011240BOS   NBA        0     2001  11/24/2000   
88233      44117  200012030BOS   NBA        0     2001   12/3/2000   
88336      44169  200012100BOS   NBA        0     2001  12/10/2000   
88469      44235  200012200BOS   NBA        0     2001  12/20/2000   
88748      44375  200101100BOS   NBA        0     2001   1/10/2001   
88925      44463  200101230BOS   NBA        0     2001   1/23/2001   
88975      44488  200101260BOS   NBA        0     2001   1/26/2001   
89013      44507  200101290BOS   NBA        0     2001   1/29/2001   
89105      44553  200102040BOS   NBA        0     2001    2/4/2001   
89141      44571  200102070BOS   NBA        0     2001    2/7/2001   
89364      44683  200102260BOS   NBA        0     2001   2/26/2001   
89425      44713  200103020BOS   NBA        0     2001    3/2/2001   
89525      44763  200103090BOS   NBA        0     2001    3/9/2001   
89756      44879  200103250BOS   NBA        0     2001   3/25/2001   
90057      45029  200104150BOS   NBA        0     2001   4/15/2001   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
87776           1            0     BOS  Celtics  103  1496.9196  1509.4957   
87805           2            0     BOS  Celtics   93  1509.4957  1512.2125   
87970           7            0     BOS  Celtics  116  1483.8416  1487.9716   
88001           8            0     BOS  Celtics  106  1487.9716  1490.5791   
88068          11            0     BOS  Celtics   96  1460.4922  1472.5193   
88093          12            0     BOS  Celtics  103  1472.5193  1477.3024   
88233          17            0     BOS  Celtics   96  1464.1198  1466.8375   
88336          20            0     BOS  Celtics  104  1450.2649  1453.4750   
88469          25            0     BOS  Celtics   99  1397.2222  1405.6613   
88748          36            0     BOS  Celtics   88  1368.3331  1384.8729   
88925          42            0     BOS  Celtics   83  1384.1893  1386.8386   
88975          43            0     BOS  Celtics   99  1386.8386  1404.1323   
89013          44            0     BOS  Celtics   99  1404.1323  1412.0132   
89105          47            0     BOS  Celtics  103  1437.8401  1448.2347   
89141          49            0     BOS  Celtics   94  1442.6997  1449.4445   
89364          57            0     BOS  Celtics   85  1468.3358  1472.3491   
89425          59            0     BOS  Celtics  103  1468.3030  1476.5731   
89525          62            0     BOS  Celtics  104  1451.1699  1459.3822   
89756          69            0     BOS  Celtics  102  1477.7058  1479.0504   
90057          81            0     BOS  Celtics   95  1463.9949  1467.3157   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87776  41.935509    DET    Pistons       83  1529.7430  1517.1669   
87805  42.220291    TOR    Raptors       91  1472.1443  1469.4275   
87970  39.678341    WAS    Wizards      109  1418.0574  1413.9274   
88001  39.951633    ATL      Hawks       99  1332.2992  1329.6917   
88068  38.062054    HOU    Rockets       81  1530.7407  1518.7136   
88093  38.561615    ORL      Magic       98  1476.9380  1472.1549   
88233  37.469780    LAC   Clippers       87  1289.0637  1286.3459   
88336  36.083122    DEN    Nuggets      102  1449.5248  1446.3147   
88469  31.235086    NJN       Nets       86  1404.6704  1396.2313   
88748  29.205793    MIA       Heat       76  1575.6382  1559.0984   
88925  29.395197    ATL      Hawks       82  1380.4136  1377.7643   
88975  31.083946    PHO       Suns       83  1546.8582  1529.5645   


In [93]:
df[
    (df['year_id']==2001) & 
    (df['fran_id']=='Celtics')&
    (df['game_location']=='H')&
    (df['game_result']=='W')
][['fran_id','pts','opp_fran', 'opp_pts']]
#seperti filter pada exel BISA ALAAAAAH!!!!
# jadi mau meliat  lawan siapa di tahun ini berapa skornya manipulasi tabelnya menyenangkan ya

fran_id  pts   opp_fran  opp_pts
87776  Celtics  103    Pistons       83
87805  Celtics   93    Raptors       91
87970  Celtics  116    Wizards      109
88001  Celtics  106      Hawks       99
88068  Celtics   96    Rockets       81
88093  Celtics  103      Magic       98
88233  Celtics   96   Clippers       87
88336  Celtics  104    Nuggets      102
88469  Celtics   99       Nets       86
88748  Celtics   88       Heat       76
88925  Celtics   83      Hawks       82
88975  Celtics   99       Suns       83
89013  Celtics   99    Pistons       87
89105  Celtics  103  Cavaliers       82
89141  Celtics   94      Bucks       91
89364  Celtics   85    Thunder       82
89425  Celtics  103       Jazz       99
89525  Celtics  104    Wizards       72
89756  Celtics  102      Bulls       98
90057  Celtics   95       Nets       88

## cara lain fungsio query

In [104]:
df.query('year_id == 2001 & fran_id == "Celtics" & game_location == "H" & game_result == "W"')[['fran_id','pts','opp_fran', 'opp_pts']]


#sebaris tapi ga rapi sekip

fran_id  pts   opp_fran  opp_pts
87776  Celtics  103    Pistons       83
87805  Celtics   93    Raptors       91
87970  Celtics  116    Wizards      109
88001  Celtics  106      Hawks       99
88068  Celtics   96    Rockets       81
88093  Celtics  103      Magic       98
88233  Celtics   96   Clippers       87
88336  Celtics  104    Nuggets      102
88469  Celtics   99       Nets       86
88748  Celtics   88       Heat       76
88925  Celtics   83      Hawks       82
88975  Celtics   99       Suns       83
89013  Celtics   99    Pistons       87
89105  Celtics  103  Cavaliers       82
89141  Celtics   94      Bucks       91
89364  Celtics   85    Thunder       82
89425  Celtics  103       Jazz       99
89525  Celtics  104    Wizards       72
89756  Celtics  102      Bulls       98
90057  Celtics   95       Nets       88

## Grouping dan aggregating

In [105]:
df.sample(2)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
108899      54450  200811300LAL   NBA        1     2009  11/30/2008   
105089      52545  200703160CHA   NBA        0     2007   3/16/2007   

        seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
108899          16            0     TOR  Raptors   99  1489.3342  1486.7302   
105089          66            0     CHA  Hornets   93  1396.1218  1384.8575   

        win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
108899  39.356178    LAL    Lakers      112  1713.7368  1716.3408   
105089  29.221699    LAC  Clippers      102  1454.6417  1465.9060   

       game_location game_result  forecast notes  
108899             A           L  0.133842   NaN  
105089             H           L  0.559413   NaN

In [106]:
gr= df.groupby('game_result')

gr

In [107]:
gr.get_group('L')

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
6               4  194611020PRO   NBA        1     1947  11/2/1946   
8               5  194611020STB   NBA        1     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126305      63153  201506070GSW   NBA        0     2015   6/7/2015   
126306      63154  201506090CLE   NBA        1     2015   6/9/2015   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Huskies   66  1300.0000  1293.2767   
3                2            0     NYK     Knicks   47  1306.7233  1297.0712   
4                1            0     DTF    Falcons   33  1300.0000  1279.6189   
6                1            0     BOS    Celtics   53  1300.0000  1294.8458   
8                1            0     PIT    Ironmen   51  1300.0000  1295.3092   
...            ...          ...     ...        ...  ...        ...        ...   
126305          99            1     GSW   Warriors   93  1806.2035  1797.5032   
126306         100            1     GSW   Warriors   91  1797.5032  1790.9591   
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
0       40.294830    NYK        Knicks       68  1300.0000  1306.7233   
3       40.692783    CHS         Stags       63  1300.0000  1309.6521   
4       38.864048    WSC      Capitols       50  1300.0000  1320.3811   
6       40.459381    PRO  Steamrollers       59  1300.0000  1305.1542   
8       40.507980    STB       Bombers       56  1300.0000  1304.6908   
...           ...    ...           ...      ...        ...        ...   
126305  67.029549    CLE     Cavaliers       95  1708.1705  1716.8708   
126306  66.615417    CLE     Cavaliers       96  1716.8708  1723.4149   
126309  60.309792    GSW      Warriors      103  1790.9591  1809.9791   
126311  60.010067    GSW      Warriors      104  1809.9791  1813.6349   
126312  59.290245    GSW      Warriors      105  1813.6349  1822.2881   

       game_location game_result  forecast notes  
0                  H           L  0.640065   NaN  
3                  A           L  0.368899   NaN  
4                  H           L  0.640065   NaN  
6                  A           L  0.359935   NaN  
8                  A           L  0.359935   NaN  
...              ...         ...       ...   ...  
126305             H           L  0.757674   NaN  
126306             A           L  0.472157   NaN  
126309             H           L  0.546572   NaN  
126311             A           L  0.234435   NaN  
126312             H           L  0.481450   NaN  

[63157 rows x 23 columns]

In [108]:
gr.get_group('W')

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
5               3  194611020DTF   NBA        1     1947  11/2/1946   
7               4  194611020PRO   NBA        0     1947  11/2/1946   
9               5  194611020STB   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126304      63153  201506070GSW   NBA        1     2015   6/7/2015   
126307      63154  201506090CLE   NBA        0     2015   6/9/2015   
126308      63155  201506110CLE   NBA        1     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id       fran_id  pts      elo_i  \
1                1            0     NYK        Knicks   68  1300.0000   
2                1            0     CHS         Stags   63  1300.0000   
5                1            0     WSC      Capitols   50  1300.0000   
7                1            0     PRO  Steamrollers   59  1300.0000   
9                1            0     STB       Bombers   56  1300.0000   
...            ...          ...     ...           ...  ...        ...   
126304          98            1     CLE     Cavaliers   95  1708.1705   
126307          99            1     CLE     Cavaliers   96  1716.8708   
126308         101            1     GSW      Warriors  103  1790.9591   
126310         102            1     GSW      Warriors  104  1809.9791   
126313         103            1     GSW      Warriors  105  1813.6349   

            elo_n  win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
1       1306.7233  41.705170    TRH    Huskies       66  1300.0000  1293.2767   
2       1309.6521  42.012257    NYK     Knicks       47  1306.7233  1297.0712   
5       1320.3811  43.135952    DTF    Falcons       33  1300.0000  1279.6189   
7       1305.1542  41.540619    BOS    Celtics       53  1300.0000  1294.8458   
9       1304.6908  41.492020    PIT    Ironmen       51  1300.0000  1295.3092   
...           ...        ...    ...        ...      ...        ...        ...   
126304  1716.8708  61.312763    GSW   Warriors       93  1806.2035  1797.5032   
126307  1723.4149  61.826408    GSW   Warriors       91  1797.5032  1790.9591   
126308  1809.9791  67.794983    CLE  Cavaliers       82  1723.4149  1704.3949   
126310  1813.6349  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391   
126313  1822.2881  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
1                  A           W  0.359935   NaN  
2                  H           W  0.631101   NaN  
5                  A           W  0.359935   NaN  
7                  H           W  0.640065   NaN  
9                  H           W  0.640065   NaN  
...              ...         ...       ...   ...  
126304             A           W  0.242326   NaN  
126307             H           W  0.527843   NaN  
126308             A           W  0.453428   NaN  
126310             H           W  0.765565   NaN  
126313             A           W  0.518550   NaN  

[63157 rows x 23 columns]

dia ngepisah yg L dan W


step 2 kita akan aggregating, siapakah itu??

In [109]:
df.groupby('game_result')['pts'].sum()

game_result
L    6148284
W    6827951
Name: pts, dtype: int64

In [110]:
df.groupby('game_result')['pts'].mean()

game_result
L     97.349209
W    108.110756
Name: pts, dtype: float64

untuk group team yang kelah memiliki agregat rata" poin 97, sedangkan untk team yang menang memiliki rata" agregat poin sebesar 108

In [111]:
df.groupby('game_result')[['fran_id','pts']].sum()

pts
game_result         
L            6148284
W            6827951

In [112]:
df.groupby('game_result')[['opp_pts','pts']].sum()

opp_pts      pts
game_result                  
L            6827951  6148284
W            6148284  6827951

In [114]:
df[
    (df['year_id']>2010) &
    (df['fran_id']=='Spurs')
].groupby(['year_id','game_result'])['game_id'].count()

year_id  game_result
2011     L              25
         W              63
2012     L              20
         W              60
2013     L              30
         W              73
2014     L              27
         W              78
2015     L              31
         W              58
Name: game_id, dtype: int64

In [118]:
df[
    (df['year_id']>2010) &
    (df['fran_id']=='Spurs')
].groupby(['year_id','game_result']).agg({'pts':'mean','game_id':'count'})

#ni menampilkan dua kolom 

pts  game_id
year_id game_result                     
2011    L             93.840000       25
        W            106.698413       63
2012    L             94.350000       20
        W            106.366667       60
2013    L             93.333333       30
        W            106.232877       73
2014    L             96.925926       27
        W            108.564103       78
2015    L             97.548387       31
        W            106.224138       58

## Manipulating Columns

-creat new coloms based on existing data
-rename columns
-drop columns

In [119]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
119371      59686  201301020TOR   NBA        1     2013   1/2/2013   
46941       23471  198203020HOU   NBA        1     1982   3/2/1982   
36198       18100  197611070SEA   NBA        1     1977  11/7/1976   
55068       27535  198603050BOS   NBA        1     1986   3/5/1986   
111018      55510  200905290DEN   NBA        0     2009  5/29/2009   

        seasongame  is_playoffs team_id       fran_id  pts      elo_i  \
119371          31            0     POR  Trailblazers   79  1470.1382   
46941           57            0     SAS         Spurs  119  1536.0278   
36198            8            0     PHI        Sixers   91  1527.3163   
55068           63            0     CHI         Bulls   97  1382.6758   
111018          98            1     DEN       Nuggets   92  1685.9653   

            elo_n  win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
119371  1457.7224  36.149582    TOR  Raptors      102  1463.2179  1475.6337   
46941   1543.1978  45.745998    HOU  Rockets      117  1558.4718  1551.3019   
36198   1520.3599  42.871517    SEA  Thunder       98  1476.4850  1483.4414   
55068   1381.5940  28.687456    BOS  Celtics      108  1739.9482  1741.0300   
111018  1661.8495  56.854748    LAL   Lakers      119  1736.0210  1760.1368   

       game_location game_result  forecast notes  
119371             A           L  0.369163   NaN  
46941              A           W  0.330739   NaN  
36198              A           L  0.429709   NaN  
55068              A           L  0.067090   NaN  
111018             H           L  0.571385   NaN

In [120]:
# create new columns based on existing data
df['selisih poin'] = df['pts'] - df['opp_pts']
df.head(5)
#melihat selisih skor

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih poin  
0  0.640065   NaN            -2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN           -16  
4  0.640065   NaN           -17

In [121]:
nba= df.copy()
#kita kopi yang baru biar kalo ada manipulasi ga mempoengaruhi yg lama

In [122]:
nba.head(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih poin  
0  0.640065   NaN            -2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN           -16  
4  0.640065   NaN           -17

In [124]:
#rename columns
nba.rename(columns={'game_location':'location','game_result':'result'}) #merubah nama kolom ni dari gameresult ke result

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Huskies   66  1300.0000  1293.2767   
1                1            0     NYK     Knicks   68  1300.0000  1306.7233   
2                1            0     CHS      Stags   63  1300.0000  1309.6521   
3                2            0     NYK     Knicks   47  1306.7233  1297.0712   
4                1            0     DTF    Falcons   33  1300.0000  1279.6189   
...            ...          ...     ...        ...  ...        ...        ...   
126309         100            1     CLE  Cavaliers   82  1723.4149  1704.3949   
126310         102            1     GSW   Warriors  104  1809.9791  1813.6349   
126311         101            1     CLE  Cavaliers   91  1704.3949  1700.7391   
126312         102            1     CLE  Cavaliers   97  1700.7391  1692.0859   
126313         103            1     GSW   Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n location  \
0       40.294830    NYK     Knicks       68  1300.0000  1306.7233        H   
1       41.705170    TRH    Huskies       66  1300.0000  1293.2767        A   
2       42.012257    NYK     Knicks       47  1306.7233  1297.0712        H   
3       40.692783    CHS      Stags       63  1300.0000  1309.6521        A   
4       38.864048    WSC   Capitols       50  1300.0000  1320.3811        H   
...           ...    ...        ...      ...        ...        ...      ...   
126309  60.309792    GSW   Warriors      103  1790.9591  1809.9791        H   
126310  68.013329    CLE  Cavaliers       91  1704.3949  1700.7391        H   
126311  60.010067    GSW   Warriors      104  1809.9791  1813.6349        A   
126312  59.290245    GSW   Warriors      105  1813.6349  1822.2881        H   
126313  68.519516    CLE  Cavaliers       97  1700.7391  1692.0859        A   

       result  forecast notes  selisih poin  
0           L  0.640065   NaN            -2  
1           W  0.359935   NaN             2  
2           W  0.631101   NaN            16  
3           L  0.368899   NaN           -16  
4           L  0.640065   NaN           -17  
...       ...       ...   ...           ...  
126309      L  0.546572   NaN           -21  
126310      W  0.765565   NaN            13  
126311      L  0.234435   NaN           -13  
126312      L  0.481450   NaN            -8  
126313      W  0.518550   NaN             8  

[126314 rows x 24 columns]

untuk menyimpan perubaahan bisa masukan ke fariabel baru, atau pakai parameter di dalamnya 

In [125]:
#rename columns
nba.rename(columns={'game_location':'location','game_result':'result'}, inplace=True) #merubah nama kolom ni dari gameresult ke result dan langsung simpan di variabel asli, atau bisa simpai tmpt lain

In [126]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n location result  forecast notes  \
0    Knicks       68  1300.0000  1306.7233        H      L  0.640065   NaN   
1   Huskies       66  1300.0000  1293.2767        A      W  0.359935   NaN   
2    Knicks       47  1306.7233  1297.0712        H      W  0.631101   NaN   
3     Stags       63  1300.0000  1309.6521        A      L  0.368899   NaN   
4  Capitols       50  1300.0000  1320.3811        H      L  0.640065   NaN   

   selisih poin  
0            -2  
1             2  
2            16  
3           -16  
4           -17

In [129]:
# drop columns
nba.drop(['elo_i','elo_n','opp_elo_i','opp_elo_n'], axis = 1, inplace = True)

#inplace = True merubah datanya langsung jadi gaperlu save di variabel

In [130]:
nba.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  pts  win_equiv opp_id opp_fran  opp_pts  \
0            0     TRH  Huskies   66   40.29483    NYK   Knicks       68   
1            0     NYK   Knicks   68   41.70517    TRH  Huskies       66   

  location result  forecast notes  selisih poin  
0        H      L  0.640065   NaN            -2  
1        A      W  0.359935   NaN             2

In [133]:
import numpy as np
nba['kolom kosong']= np.NaN
nba

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts  win_equiv opp_id  \
0                1            0     TRH    Huskies   66  40.294830    NYK   
1                1            0     NYK     Knicks   68  41.705170    TRH   
2                1            0     CHS      Stags   63  42.012257    NYK   
3                2            0     NYK     Knicks   47  40.692783    CHS   
4                1            0     DTF    Falcons   33  38.864048    WSC   
...            ...          ...     ...        ...  ...        ...    ...   
126309         100            1     CLE  Cavaliers   82  60.309792    GSW   
126310         102            1     GSW   Warriors  104  68.013329    CLE   
126311         101            1     CLE  Cavaliers   91  60.010067    GSW   
126312         102            1     CLE  Cavaliers   97  59.290245    GSW   
126313         103            1     GSW   Warriors  105  68.519516    CLE   

         opp_fran  opp_pts location result  forecast notes  selisih poin  \
0          Knicks       68        H      L  0.640065   NaN            -2   
1         Huskies       66        A      W  0.359935   NaN             2   
2          Knicks       47        H      W  0.631101   NaN            16   
3           Stags       63        A      L  0.368899   NaN           -16   
4        Capitols       50        H      L  0.640065   NaN           -17   
...           ...      ...      ...    ...       ...   ...           ...   
126309   Warriors      103        H      L  0.546572   NaN           -21   
126310  Cavaliers       91        H      W  0.765565   NaN            13   
126311   Warriors      104        A      L  0.234435   NaN           -13   
126312   Warriors      105        H      L  0.481450   NaN            -8   
126313  Cavaliers       97        A      W  0.518550   NaN             8   

        kolom kosong  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
...              ...  
126309           NaN  
126310           NaN  
126311           NaN  
126312           NaN  
126313           NaN  

[126314 rows x 21 columns]